### PR TITLE
qemu.tests.nic_hotplug: add missed reboot step

### PR DESCRIPTION
**Pull request by xutian**
1. add steps reboot vm after nic hotplug and hotunplug
2. check vm status after pause & resume
3. add debug info when raise VMAddressVerificationError and VMIPAddressMissingError

Signed-off-by: Xu Tian <xutian@redhat.com>

### CODE REVIEW

**Comment by xutian** [comment]
ID: 1128020

**Comment by xutian** [comment]
deps on https://github.com/autotest/virt-test/pull/1973

**Comment by xutian** [comment]
deps on https://github.com/autotest/virt-test/pull/1985

File: qemu/tests/nic_hotplug.py [hunk 1]
```diff
@@ -50,6 +49,9 @@ def run(test, params, env):
     nic_hotplugged = []
     try:
         for nic_index in xrange(1, nic_hotplug_count + 1):
+            # need to reconnect serial port after
+            # guest reboot for windows guest
+            session = vm.wait_for_serial_login(timeout=login_timeout)
```
Comment (CongSmile) [inline qemu/tests/nic_hotplug.py]: Hi @xutian 

Do we need judge the guest which is windows or linux here? 

Thanks.

File: qemu/tests/nic_hotplug.py [hunk 2]
```diff
@@ -99,17 +119,31 @@ def run(test, params, env):
...
+            s_session = vm.reboot(session=s_session, serial=True)
+            vm.verify_alive()
+            if guest_is_linux:
+                s_session.cmd("dhclient -r", timeout=120)
+                s_session.cmd("dhclient %s" % ifname, timeout=240)
+            else:
+                s_session.cmd("ipconfig /renew all", timeout=240)
+            hotnic_ip = vm.wait_for_get_address(nic_name)
+
+            error.context("Ping guest's new ip after resume", logging.info)
```
Comment (CongSmile) [inline qemu/tests/nic_hotplug.py]: Seems this context should be behind "cont".

**Comment by xiagao** [comment]
failed with win7 guest.fail log as follows:
01/28 15:34:03 INFO |     error:0085| Context: Check if new interface gets ip address
01/28 15:34:04 DEBUG|     utils:0116| Running '/sbin/arping -f -c 5 -I switch 10.66.109.217'
01/28 15:34:04 DEBUG|     utils:0153| [stdout] ARPING 10.66.109.217 from 10.66.106.5 switch
01/28 15:34:04 DEBUG|     utils:0153| [stdout] Unicast reply from 10.66.109.217 [2C:44:FD:7E:51:68]  0.950ms
01/28 15:34:04 DEBUG|     utils:0153| [stdout] Sent 1 probes (1 broadcast(s))
01/28 15:34:04 DEBUG|     utils:0153| [stdout] Received 1 response(s)
01/28 15:34:04 DEBUG|     utils:0116| Running '/sbin/arping -f -c 5 -I switch 10.66.106.5'
01/28 15:34:07 DEBUG|qemu_monit:0267| (monitor qmpmonitor1) Sending command 'info registers' (via Human Monitor)
01/28 15:34:09 DEBUG|     utils:0153| [stdout] ARPING 10.66.106.5 from 10.66.106.5 switch
01/28 15:34:09 DEBUG|     utils:0153| [stdout] Sent 5 probes (5 broadcast(s))
01/28 15:34:09 DEBUG|     utils:0153| [stdout] Received 0 response(s)
01/28 15:34:09 DEBUG|   virt_vm:0675| Found/Verified IP 10.66.107.9 for VM virt-tests-vm1 NIC hotplug_nic3
01/28 15:34:09 DEBUG|   virt_vm:0675| Found/Verified IP 10.66.107.9 for VM virt-tests-vm1 NIC hotplug_nic3
01/28 15:34:09 INFO |nic_hotplu:0115| Got the ip address of new nic: 10.66.107.9
01/28 15:34:09 INFO |     error:0085| Context: Ping guest's new ip from host
01/28 15:34:09 INFO |  __init__:0987| The command of Ping is: ping 10.66.107.9  -c 10
01/28 15:34:09 DEBUG|   aexpect:0965| PING 10.66.107.9 (10.66.107.9) 56(84) bytes of data.
01/28 15:34:13 DEBUG|qemu_monit:0267| (monitor qmpmonitor1) Sending command 'info registers' (via Human Monitor)
01/28 15:34:18 DEBUG|qemu_monit:0267| (monitor qmpmonitor1) Sending command 'info registers' (via Human Monitor)
01/28 15:34:19 INFO | utils_env:0092| tcpdump updated address_cache(9a:a0:a1:a2:a3:a4 <-> 10.66.107.138)
01/28 15:34:19 INFO | utils_env:0092| tcpdump updated address_cache(9a:a0:a1:a2:a3:a4 <-> 10.66.107.138)
01/28 15:34:21 DEBUG|   aexpect:0965| From 10.66.106.5 icmp_seq=10 Destination Host Unreachable

check arp table in host:
# arp -ane |grep 10.66.107.9 
10.66.107.9                      (incomplete)                              switch

**Comment by xutian** [comment]
On 01/28/2015 04:36 PM, xiagao wrote:
>
> failed with win7 guest.fail log as follows:
> 01/28 15:34:03 INFO | error:0085| Context: Check if new interface gets 
> ip address
> 01/28 15:34:04 DEBUG| utils:0116| Running '/sbin/arping -f -c 5 -I 
> switch 10.66.109.217'
> 01/28 15:34:04 DEBUG| utils:0153| [stdout] ARPING 10.66.109.217 from 
> 10.66.106.5 switch
> 01/28 15:34:04 DEBUG| utils:0153| [stdout] Unicast reply from 
> 10.66.109.217 [2C:44:FD:7E:51:68] 0.950ms
> 01/28 15:34:04 DEBUG| utils:0153| [stdout] Sent 1 probes (1 broadcast(s))
> 01/28 15:34:04 DEBUG| utils:0153| [stdout] Received 1 response(s)
> 01/28 15:34:04 DEBUG| utils:0116| Running '/sbin/arping -f -c 5 -I 
> switch 10.66.106.5'
> 01/28 15:34:07 DEBUG|qemu_monit:0267| (monitor qmpmonitor1) Sending 
> command 'info registers' (via Human Monitor)
> 01/28 15:34:09 DEBUG| utils:0153| [stdout] ARPING 10.66.106.5 from 
> 10.66.106.5 switch
> 01/28 15:34:09 DEBUG| utils:0153| [stdout] Sent 5 probes (5 broadcast(s))
> 01/28 15:34:09 DEBUG| utils:0153| [stdout] Received 0 response(s)
> 01/28 15:34:09 DEBUG| virt_vm:0675| Found/Verified IP 10.66.107.9 for 
> VM virt-tests-vm1 NIC hotplug_nic3
> 01/28 15:34:09 DEBUG| virt_vm:0675| Found/Verified IP 10.66.107.9 for 
> VM virt-tests-vm1 NIC hotplug_nic3
> 01/28 15:34:09 INFO |nic_hotplu:0115| Got the ip address of new nic: 
> 10.66.107.9
> 01/28 15:34:09 INFO | error:0085| Context: Ping guest's new ip from host
> 01/28 15:34:09 INFO | *init*:0987| The command of Ping is: ping 
> 10.66.107.9 -c 10
> 01/28 15:34:09 DEBUG| aexpect:0965| PING 10.66.107.9 (10.66.107.9) 
> 56(84) bytes of data.
> 01/28 15:34:13 DEBUG|qemu_monit:0267| (monitor qmpmonitor1) Sending 
> command 'info registers' (via Human Monitor)
> 01/28 15:34:18 DEBUG|qemu_monit:0267| (monitor qmpmonitor1) Sending 
> command 'info registers' (via Human Monitor)
> 01/28 15:34:19 INFO | utils_env:0092| tcpdump updated 
> address_cache(9a:a0:a1:a2:a3:a4 <-> 10.66.107.138)
> 01/28 15:34:19 INFO | utils_env:0092| tcpdump updated 
> address_cache(9a:a0:a1:a2:a3:a4 <-> 10.66.107.138)
> 01/28 15:34:21 DEBUG| aexpect:0965| From 10.66.106.5 icmp_seq=10 
> Destination Host Unreachable
>
> check arp table in host:
>
>
>   arp -ane |grep 10.66.107.9
>
> 10.66.107.9 (incomplete) switch
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/autotest/tp-qemu/pull/259#issuecomment-71798315>.
>
Hi gaoxi,

Do you check is guest really use IP 10.66.107.9

File: qemu/tests/nic_hotplug.py (hunk 1 above)
Comment (xutian) [inline qemu/tests/nic_hotplug.py]: yes, we need as commented in above line that we have reboot steps in for loop, for windows guest serial_session will lost when guest reboot.

Thanks,
@xutian

File: qemu/tests/nic_hotplug.py [hunk 3]
```diff
@@ -99,17 +119,31 @@ def run(test, params, env):
...
+            # reboot vm via serial port since some guest can't auto up
+            # hotplug nic and next step will check is hotplug nic works.
+            s_session = vm.reboot(session=s_session, serial=True)
+            vm.verify_alive()
+            if guest_is_linux:
+                s_session.cmd("dhclient -r", timeout=120)
+                s_session.cmd("dhclient %s" % ifname, timeout=240)
+            else:
+                s_session.cmd("ipconfig /renew all", timeout=240)
+            hotnic_ip = vm.wait_for_get_address(nic_name)
```
Comment (xiagao) [inline qemu/tests/nic_hotplug.py]: Hi,@xutian
1.suggest add context with  "error.context("Check if new interface gets ip address after reboot", logging.info)"
2.wait_for_get_address(nic_name) should add timeout.
3.if failed to get guest's address,please provide error message such as "err_msg = "Could not get or verify nic ip address""
4.if failed to get guest's address ,should add check code such as line 95-116.

thanks,
xiagao

**Comment by xutian** [comment]
deps on https://github.com/autotest/virt-test/pull/1988  & https://github.com/autotest/virt-test/pull/1985

**Comment by xutian** [comment]
@xiagao  & @CongSmile 

could you help to review it again.

thanks,
@xutian

**Comment by CongSmile** [comment]
Acked-by: Cong Li <coli@redhat.com>

**Comment by xiagao** [comment]
Acked-by:Xiaoling Gao xiagao@redhat.com

**Comment by FengYang** [comment]
Thanks @xutian , @xiagao @CongSmile

**Pull request merged**